### PR TITLE
[5.1] Add Global Setting for Form Layout option to custom fields

### PR DESCRIPTION
**Pull request by AndySDH**
Pull Request for Issue #37162. Follow-up to PR #36551 and issue #36548.

### Purpose of the PR

As discussed in #37162, this allows users to set a global preference for the HTML Select for custom fields of types: list, SQL, radio, integer.

I would have just removed the option, as the "raw" select is just ugly, but @laoneo noted that we need to keep it for Backwards Compatibility.

### Summary of Changes

- Added plugin settings for all relevant fields so that the field inherits from the global options by default (this allows users to set a global preference instead of having to set it for each field they create)
- The global option will still use the old "raw" style by default (to keep Backwards Compatibility)
- Implemented the same setting for the Integer field type as well (it was forgotten about in PR 36551 )
- Renamed the option to "Form Layout" for it to be more explicative
- Added new language strings in a shared language file to avoid duplicating them for all field types

### Testing Instructions

First create the custom field:
- Create Custom Field of type 'list' for the Content component. Make sure that the 'Multiple' setting is set to 'Yes'
- Go to Options->Form Options: Keep the 'Form Layout' as 'Use Settings from Plugin'

Then:
- Go to Plugin Manager, and in the plugin options for 'Fields - List', change the default 'Form Layout' option
- Edit an article and check the form display of the field, it will change based on your selection of 'Form Layout'

### Actual result BEFORE applying this Pull Request

There was no global setting to inherit from the plugin for 'Form Layout'

### Expected result AFTER applying this Pull Request

There is now a global setting for 'Form Layout' that controls the default value when not specified in an individual field.

Changing the global setting will change the form layout of the field.

### CODE REVIEW

File: administrator/language/en-GB/com_fields.ini [hunk 1]
```diff
@@ -21,7 +21,9 @@ COM_FIELDS_FIELD_EDITABLE_IN_BOTH="Both"
 COM_FIELDS_FIELD_EDITABLE_IN_LABEL="Editable In"
 COM_FIELDS_FIELD_EDITABLE_IN_SITE="Site"
 COM_FIELDS_FIELD_FORM_EDIT="Edit Field"
-COM_FIELDS_FIELD_FORM_LAYOUT_LABEL="Layout"
+COM_FIELDS_FIELD_FORM_LAYOUT_LABEL="Form Layout"
+COM_FIELDS_FIELD_FORM_LAYOUT_BASIC_SELECT="HTML Select"
+COM_FIELDS_FIELD_FORM_LAYOUT_FANCY_SELECT="Enhanced Select"
```
Comment (brianteeman) [inline administrator/language/en-GB/com_fields.ini]:
```suggestion
COM_FIELDS_FIELD_FORM_LAYOUT_BASIC_SELECT="HTML Select"
COM_FIELDS_FIELD_FORM_LAYOUT_FANCY_SELECT="Enhanced Select"
COM_FIELDS_FIELD_FORM_LAYOUT_LABEL="Form Layout"
```

Accept this to order the strings alphabetically

File: plugins/fields/integer/params/integer.xml [hunk 2]
```diff
@@ -13,6 +13,17 @@
 				<option value="1">JYES</option>
 				<option value="0">JNO</option>
 			</field>
+			
+			<field
+				name="form_layout"
+				type="list"
+				label="COM_FIELDS_FIELD_FORM_LAYOUT_LABEL"
+				class="form-select"
```
Comment (brianteeman) [inline plugins/fields/integer/params/integer.xml]:
```suggestion
				useglobal="true"
				class="form-select"
```

File: plugins/fields/list/params/list.xml [hunk 3]
```diff
@@ -13,6 +13,17 @@
 				<option value="1">JYES</option>
 				<option value="0">JNO</option>
 			</field>
+			
+			<field
+				name="form_layout"
+				type="list"
+				label="COM_FIELDS_FIELD_FORM_LAYOUT_LABEL"
+				class="form-select"
```
Comment (brianteeman) [inline plugins/fields/list/params/list.xml]:
```suggestion
				useglobal="true"
				class="form-select"
```

File: plugins/fields/radio/params/radio.xml [hunk 4]
```diff
@@ -2,6 +2,17 @@
 <form>
 	<fields name="fieldparams">
 		<fieldset name="fieldparams">
+			<field
+				name="form_layout"
+				type="list"
+				label="COM_FIELDS_FIELD_FORM_LAYOUT_LABEL"
+				class="form-select"
```
Comment (brianteeman) [inline plugins/fields/radio/params/radio.xml]:
```suggestion
				useglobal="true"
				class="form-select"
```

File: plugins/fields/sql/params/sql.xml [hunk 5]
```diff
@@ -23,21 +23,17 @@
...
-					<option value="joomla.form.field.list">JDEFAULT</option>
-					<option value="joomla.form.field.list-fancy-select">PLG_FIELDS_SQL_PARAMS_FORM_LAYOUT_FANCY_SELECT</option>
-				</field>
-			</fieldset>
+			
+			<field
+				name="form_layout"
+				type="list"
+				label="COM_FIELDS_FIELD_FORM_LAYOUT_LABEL"
+				class="form-select"
```
Comment (brianteeman) [inline plugins/fields/sql/params/sql.xml]:
```suggestion
				useglobal="true"
				class="form-select"
```

**Comment by AndySDH** [comment]
Thanks @brianteeman

File: administrator/components/com_fields/src/Plugin/FieldsPlugin.php [hunk 6]
```diff
@@ -210,6 +205,13 @@ public function onCustomFieldsPrepareDom($field, \DOMElement $parent, Form $form
 		$params = clone $this->params;
 		$params->merge($field->fieldparams);
 
+		$layout = $params->get('form_layout');
```
Comment (laoneo) [inline administrator/components/com_fields/src/Plugin/FieldsPlugin.php]: Why this change?

File: administrator/components/com_plugins/src/View/Plugin/HtmlView.php [hunk 7]
```diff
@@ -66,6 +66,12 @@ public function display($tpl = null)
 			throw new GenericDataException(implode("\n", $errors), 500);
 		}
 
+		// Support for Language Strings from com_fields
```
Comment (laoneo) [inline administrator/components/com_plugins/src/View/Plugin/HtmlView.php]: Why is this needed? Till now we could solve it in the respective plugins. I would like to avoid here a hard dependency to com_fields.

Comment (AndySDH) [reply administrator/components/com_plugins/src/View/Plugin/HtmlView.php]: I did that so we can use these shared language strings across multiple field plugins. This is to avoid duplicating the 3 related language strings for each plugin. Do you have a better way in which we could to that?

File: administrator/components/com_fields/src/Plugin/FieldsPlugin.php (hunk 6 above)
Comment (AndySDH) [reply administrator/components/com_fields/src/Plugin/FieldsPlugin.php]: PHP 8 will throw a warning when setting a value inside an if tag. It's better to set it outside of the if tag, and then check against it.

File: administrator/components/com_plugins/src/View/Plugin/HtmlView.php (hunk 7 above)
Comment (AndySDH) [reply administrator/components/com_plugins/src/View/Plugin/HtmlView.php]: I'm fixing it by moving the language strings to lib_joomla.ini

**Comment by AndySDH** [comment]
Updated PR, please test it now guys, thanks! :)

File: language/en-GB/lib_joomla.ini [hunk 8]
```diff
@@ -297,6 +297,9 @@ JLIB_FORM_FIELD_PARAM_INTEGER_LAST_DESC="This value is the highest in the list."
 JLIB_FORM_FIELD_PARAM_INTEGER_LAST_LABEL="Last"
 JLIB_FORM_FIELD_PARAM_INTEGER_STEP_DESC="Each option will be the previous option incremented by this integer, starting with the first value until the last value is reached."
 JLIB_FORM_FIELD_PARAM_INTEGER_STEP_LABEL="Step"
+JLIB_FORM_FIELD_PARAM_LAYOUT_BASIC_SELECT="HTML Select"
```
Comment (brianteeman) [inline language/en-GB/lib_joomla.ini]: Dont these need to be in the admin language file ?

Comment (AndySDH) [reply language/en-GB/lib_joomla.ini]: I moved them to the lib_joomla.ini file so that they can be shared across multiple fields without duplicating them in different language files. If you have a better exact file where these should be placed, let me know, as long as that file is loaded by com_plugins AND com_fields automatically

Comment (regularlabs) [reply language/en-GB/lib_joomla.ini]: This is also the language file where other - even more specific - field plugin strings are stored, like:
```
JLIB_FORM_FIELD_PARAM_CHECKBOX_MULTIPLE_LABEL="Multiple"
JLIB_FORM_FIELD_PARAM_LIST_MULTIPLE_LABEL="Multiple"
JLIB_FORM_FIELD_PARAM_RADIO_MULTIPLE_LABEL="Multiple"
```

No idea why there isn't just 1 general string for that, like:
```
JLIB_FORM_FIELD_PARAM_MULTIPLE_LABEL="Multiple"
```

**Comment by AndySDH** [comment]
Can you guys please test this? @brianteeman @laoneo @regularlabs

**Comment by laoneo** [comment]
Can you rebase it to the 4.2-dev branch as it is a new feature?

**Comment by AndySDH** [comment]
Not sure how to do that, I create PRs by manually editing files here on github... so I'm not sure how to rebase this against a different branch. 

Is there a way to do this without having to re-create it from scratch? Can someone with more knowledge/permissions do this?

**Comment by AndySDH** [comment]
Done, I was able to do it! @laoneo 
Thanks to the docs haha: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-base-branch-of-a-pull-request

**Comment by brianteeman** [comment]
https://user-images.githubusercontent.com/1296369/159136063-8470361b-9b14-4cd5-9654-d4bb58cd143d.gif

**Comment by laoneo** [comment]
When the next time the branch is merged from 4.1-dev into 4.2-dev, then the commits which are not from you will disappear.

File: plugins/fields/sql/sql.xml [hunk 9]
```diff
@@ -28,20 +28,31 @@
 					description="PLG_FIELDS_SQL_PARAMS_QUERY_DESC"
 					rows="10"
 					filter="raw"
-					required="true"
```
Comment (laoneo) [inline plugins/fields/sql/sql.xml]: Please make a new pr with this change. Would be good to figure out why this needed, otherwise it is hard to trace back. Thanks for understanding.

Comment (AndySDH) [reply plugins/fields/sql/sql.xml]: I explained why it's needed. It doesn't make sense for a global "default" setting to be "required". Otherwise you can't save the plugin settings at all unless you fill that in. So currently if you want to just change the default "Multiple" setting, you can't.

It's not hard to trace back. I explained it in the PR. It's such a minor thing that affects no one, other than the inability to save the plugin settings before.

Comment (laoneo) [reply plugins/fields/sql/sql.xml]: Look, I don't want to discus this in here. It belongs to it's own pr as I do not agree with this change. Better not to block this one here out, or?

Comment (AndySDH) [reply plugins/fields/sql/sql.xml:31]: So, explain, how do you not agree with this change? You think you shouldn't be able to save the plugin settings unless you're forced to give a default query? Why would you want to have a default query? That makes no sense.

But whatever, I'll remove it. I won't bother creating a new PR for it, though. You can keep the bugged version

Comment (AndySDH) [reply plugins/fields/sql/sql.xml]: If after making me jump through hoops to change this, you now have at least the courtesy to mark it as tested so we can move forward, it would be appreciated! ;)

**Comment by laoneo** [comment]
When I want to create an integer field there are untranslated strings:
image

**Comment by AndySDH** [comment]
Thank you for spotting this, sorry, please try now.

**Comment by laoneo** [comment]
I have tested this item :white_check_mark: successfully on 9137eee3df49ab57d87d772f810afd922d71dc0a<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37320">issues.joomla.org/tracker/joomla-cms/37320</a>.</sub>

File: plugins/fields/integer/integer.xml [hunk 10]
```diff
@@ -24,7 +24,7 @@
 				<field
 					name="multiple"
 					type="radio"
-					label="PLG_FIELDS_INTEGER_PARAMS_MULTIPLE_LABEL"
```
Comment (Quy) [inline plugins/fields/integer/integer.xml]: Delete language string? Should this be a separate PR for similar changes?

Comment (AndySDH) [reply plugins/fields/integer/integer.xml]: Makes sense, I'll revert the changes to 'multiple' string for now and tackle them in another PR if needed.

**Comment by AndySDH** [comment]
Please test again, thanks guys!

**Comment by Quy** [comment]
`Use Global` and `Use settings from Plugin` have the same value.

form-layout

```
<select id="jform_fieldparams_form_layout" name="jform[fieldparams][form_layout]" class="form-select form-select">
	<option value="" selected="selected">Use Global</option>
	<option value="" selected="selected">Use settings from Plugin</option>
	<option value="joomla.form.field.list">HTML Select</option>
	<option value="joomla.form.field.list-fancy-select">Enhanced Select</option>
</select>
```

**Comment by Quy** [comment]
`Use Global` and `Use settings from Plugin` have the same value.

form-layout

```
<select id="jform_fieldparams_form_layout" name="jform[fieldparams][form_layout]" class="form-select form-select">
	<option value="" selected="selected">Use Global</option>
	<option value="" selected="selected">Use settings from Plugin</option>
	<option value="joomla.form.field.list">HTML Select</option>
	<option value="joomla.form.field.list-fancy-select">Enhanced Select</option>
</select>
```

**Comment by AndySDH** [comment]
Thanks for spotting this @Quy, I guess this was due to @brianteeman suggestion of adding useglobal="true".

What is the standard practice here? Should we remove the `useglobal="true"` attribute, or should we remove `<option value="">COM_FIELDS_FIELD_USE_GLOBAL</option>` ?

I see that, for comparison, the 'multiple' field is not using `useglobal="true"` as an attribute.

**Comment by brianteeman** [comment]
useglobal = true should result in the select showing something like "Use Gobal (HTML Select)" which you cannot see when you have it as an option

**Comment by AndySDH** [comment]
That's interesting. Though It doesn't seem to happen from @Quy screen for some reason it seems, @Quy can you confirm?

If it works that way, we should probably also do an additional PR to cover other cases where useglobal="true" is not yet used (it doesn't seem to be implemented everywhere, maybe not for field plugins) @brianteeman

**Comment by Quy** [comment]
It does not display as seen in the screenshot. I would say do a separate PR for `useglobal` and not include it in this PR if it is available in plugins.

**Comment by AndySDH** [comment]
Thanks for the feedback. I removed useglobal="true" for now.

Please test again! Thanks!

**Comment by AndySDH** [comment]
Hey guys, any update on this?

**Comment by AndySDH** [comment]
Hey guys, any update on this?

**Comment by HLeithner** [comment]
This pull request has been automatically rebased to 5.0-dev.

**Comment by AndySDH** [comment]
> This pull request has been automatically rebased to 5.0-dev.

So I guess this PR is scheduled for Joomla 5?

**Comment by crommie** [comment]
After applying the patch I get this error: Class "Joomla\Plugin\Fields\Integer\Extension\Integer" not found

**Comment by flo-the-cat** [comment]
I have tested this item :red_circle: unsuccessfully on 9137eee3df49ab57d87d772f810afd922d71dc0a<br /><br />Althought the patch does indeed add the required field, you cannot then use ti as adding an article is broken with the following message
 Class "Joomla\Plugin\Fields\Integer\Extension\Integer" not found

Reverting the patch and you can add articles again <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37320">issues.joomla.org/tracker/joomla-cms/37320</a>.</sub>

**Comment by oshufx** [comment]
I have tested this item :red_circle: unsuccessfully on 9137eee3df49ab57d87d772f810afd922d71dc0a<br /><br />This test was unsuccessful due to articles no longer being able to be created after the patch had been applied. This is the error that returned.

An error has occurred.
0 Class "Joomla\Plugin\Fields\Integer\Extension\Integer" not found

<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37320">issues.joomla.org/tracker/joomla-cms/37320</a>.</sub>

**Comment by richard67** [comment]
@crommie @flo-the-cat @oshufx The reason for the unsuccessful tests is because this PR has merge conflicts. You can see them when going to the PR and scroll to the bottom. I will try to solve them.

**Comment by flo-the-cat** [comment]
> @crommie @flo-the-cat @oshufx The reason for the unsuccessful tests is because this PR has merge conflicts. You can see them when going to the PR and scroll to the bottom. I will try to solve them.

Just say when and we will retest

**Comment by richard67** [comment]
> Just say when and we will retest

@crommie @flo-the-cat @oshufx Ready. If you use Patchtester, just don't forget to revert this and any other PRs and then fetching the list of PRs again before applying this PR again.

**Comment by crommie** [comment]
I have tested this item :white_check_mark: successfully on 9137eee3df49ab57d87d772f810afd922d71dc0a<br /><br />Without patch:

Form layout (in options tab) has options Default and Enhanced Select
Plugin has option to enable/disable Multiple and add a Default Value for the list

With patch:

Form layout is now in the first tab and no longer in the options tab. It has options Use settings from plugin, HTML Select and Enhanced Select
Plugin has option to enable/disable Multiple, add a Default Value for the list and a dropdown to choose HTML Select or Enhanced Select
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37320">issues.joomla.org/tracker/joomla-cms/37320</a>.</sub>

**Comment by AndySDH** [comment]
Thanks for testing! One more test to merge :)

**Comment by AndySDH** [comment]
I have decided to move back the form_layout setting under params instead of fieldparams. For two reasons:
- For backwards compatibility. Moving it to fieldparams would mean saved values would have been lost
- It's under params for the subform, and the option is named the same. So it's best if it's consistent across all field types.

 So the option has now returned once again under "Options -> Form Options".

Please re-test again, thank you :)

**Comment by ceford** [comment]
I am afraid the test instruction are not clear enough for me to follow:
`Go to Plugin Manager, and in the plugin options of 'Fields - List', change the default 'Form Layout' option`
Change it to what? I have a choice of HTML Select or Enhanced Select. I created some list values here too - but that does not seem right and they are not used.

`Edit an article and check the form display of the field`
I get either an empty list box or an empty textarea. I don't know what I am looking for.

Testing in 5.0.0-beta2-dev<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37320">issues.joomla.org/tracker/joomla-cms/37320</a>.</sub>

**Comment by AndySDH** [comment]
> I am afraid the test instruction are not clear enough for me to follow: `Go to Plugin Manager, and in the plugin options of 'Fields - List', change the default 'Form Layout' option` Change it to what? I have a choice of HTML Select or Enhanced Select. I created some list values here too - but that does not seem right and they are not used.

@ceford 

Set it to "Enhanced Select" for example. When editing the field in the article, you should see this:

image

While if you set it to "HTML Select" you should see this:

image

**Comment by ceford** [comment]
There are problems with this PR. It adds a form field to four Field Plugins: Integer, List, Radio and SQL. But the radio field has a Buttons/Switcher choice and the SQL plugin cannot be saved unless a default SQL query is entered. 

Applying the patch and creating a Field of each type, when I edit an article the Fields tab gives me the following errors, before the first field:
```
Warning
: Undefined property: stdClass::$value in
/Users/ceford/Sites/joomla-cms/libraries/src/Form/Field/SqlField.php
on line
302


Warning
: Undefined property: stdClass::$text in
/Users/ceford/Sites/joomla-cms/libraries/src/Form/Field/SqlField.php
on line
302


Deprecated
: trim(): Passing null to parameter #1 ($string) of type string is deprecated in
/Users/ceford/Sites/joomla-cms/libraries/src/HTML/Helpers/Select.php
on line
388
```
For the SQL field:
```
 Deprecated: explode(): Passing null to parameter #2 ($string) of type string is deprecated in /Users/ceford/Sites/joomla-cms/libraries/src/HTML/Helpers/Select.php on line 557
```
Could be my fault for not filling out the fields properly.

**Comment by AndySDH** [comment]
Thanks for testing @ceford 

> the SQL plugin cannot be saved unless a default SQL query is entered.

That is a separate issue not coming from this PR. It was an issue already present. I tried to fix it as part of this PR as well, but @laoneo asked me to remove it from this PR: https://github.com/joomla/joomla-cms/pull/37320#discussion_r836175947

So that is a separate issue. If it was me I would have included it here, but I didn't make the decision.

> Applying the patch and creating a Field of each type, when I edit an article the Fields tab gives me the following errors, before the first field:

I don't think those warnings are coming from this PR. I didn't touch those files, and none of the changes I made should affect those files.

File: administrator/components/com_fields/src/Plugin/FieldsPlugin.php [hunk 11]
```diff
@@ -282,6 +278,12 @@ public function onCustomFieldsPrepareDom($field, \DOMElement $parent, Form $form
         $params = clone $this->params;
         $params->merge($field->fieldparams);
 
+        $layout = $params->get('form_layout');
```
Comment (obuisard) [inline administrator/components/com_fields/src/Plugin/FieldsPlugin.php]:
```suggestion
        $layout = $field->params->get('form_layout', $this->params->get('form_layout', ''));
```
The layout is not found in $params.
Added the value from the plugin as default.

**Comment by obuisard** [comment]
I tested every single field. None of them is using the global settings from their associated plugin.
I think you may have broken that part when you reverted the changes (from fieldparams to where it is now).

**Comment by AndySDH** [comment]
@obuisard That is extremely weird. After applying the patch, all fields inherit the layout from the respective global settings. I just tested it again to make sure and still worked for me. So I'm not sure why it's not working on your end.

I just applied your proposed change though. Can you please test again now? Thank you!

**Comment by obuisard** [comment]
> @obuisard That is extremely weird. After applying the patch, for me all fields inherit the layout from the respective global settings. I just tested it again to make sure and still worked for me. So I'm not sure why it's not working on your end.

Could be a side effect if you have not saved the fields again after the location change of the form_layout.

**Comment by AndySDH** [comment]
Thank you :)

**Comment by HLeithner** [comment]
This pull request has been automatically rebased to 5.1-dev.

**Comment by AndySDH** [comment]
Needs one more test please :)

**Comment by ceford** [comment]
I have tested this and have not triggered any errors this time. I think it is OK but do not know how to create a ticked pass box as I usually do that with the Joomla test interface which is currently down.

**Comment by obuisard** [comment]
> I have tested this and have not triggered any errors this time. I think it is OK but do not know how to create a ticked pass box as I usually do that with the Joomla test interface which is currently down.

You will have to wait for the issue tracker to get back online, I am afraid. This is the only way you can cast your test. In the meantime, Richard set the PR as RTC, which acknowledges your test.

**Comment by AndySDH** [comment]
> I have tested this and have not triggered any errors this time. I think it is OK but do not know how to create a ticked pass box as I usually do that with the Joomla test interface which is currently down.

Thank you :)

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37320">issues.joomla.org/tracker/joomla-cms/37320</a>.</sub>

**Comment by Razzo1987** [comment]
I have tested this item :white_check_mark: successfully on 9137eee3df49ab57d87d772f810afd922d71dc0a<br /><br />Very nice work! We love it!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37320">issues.joomla.org/tracker/joomla-cms/37320</a>.</sub>
